File: Puerta.cpp
// Archivo Puerta.cpp

#include "Puerta.h"

// Constructor
Puerta::Puerta(int id, string ubicacion, string disponibilidad)
    : Id(id), Ubicacion(ubicacion), Disponibilidad(disponibilidad){
        act = nullptr;
        Horas hora(0, 0);
        Hora_Embarque = hora;
        
    }

int Puerta::getId() {
    return Id;
}

void Puerta::setId(int newId) {
    Id = newId;
}

string Puerta::getUbicacion() {
    return Ubicacion;
}

void Puerta::setUbicacion(string newUbicacion) {
    Ubicacion = newUbicacion;
}

string Puerta::getDisponibilidad() {
    return Disponibilidad;
}

void Puerta::setDisponibilidad(string newDisponibilidad) {
    Disponibilidad = newDisponibilidad;
}

Horas Puerta::getHora_Embarque() {
    return Hora_Embarque;
}

void Puerta::setHoraDeEmbarque(Horas hora) {
    Hora_Embarque = hora;
}

void Puerta::getHistorial() {
    // Implementación de la función para obtener historial
    int tam = historial.size(), Info = 0, vuelo_Sel = 0;
    if(tam != 0){
        if(tam > 1){
            cout << "Los vuelos que han sido asignados a esta puerta son:" << endl;
            for(int i = 0; i < tam; i++){
                cout << i + 1 << ". " << historial[i]->getNo_Identificacion() << "." << endl;
            }
            cout << "Desea conocer mas informacion de alguno de los vuelos? \n1. Si.\n2. No.";
            switch(Info){
                case 1:
                    cout << "Sobre que vuelo desea recibir informacion? (Indique el numero del indice)" << endl;
                    cin >> vuelo_Sel;
                    if(vuelo_Sel > 0 && vuelo_Sel < tam){
                        historial[vuelo_Sel - 1]->Show_Info(true);
                    }
                    break;
                case 2:
                    cout << "Entendido" << endl;
                    break;
                default:
                    cout << "Valor indefinido" << endl;
                    break;
            }
        }
        else{
            cout << "El unico vuelo que ha sido asignado a esta puerta es el vuelo: " << historial[0]->getNo_Identificacion() << "\nDesea conocer mas informacion? \n1. Si.\n2. No." << endl;
            cin >> Info;
            switch(Info){
                case 1:
                    historial[0]->Show_Info(true);
                    break;
                case 2:
                    cout << "Entendido" << endl;
                    break;
                default:
                    cout << "Valor indefinido" << endl;
                    break;
            }
        }
    }
    else{
        cout << "No se ha asignado ningun vuelo a esta puerta" << endl;
    }
}

void Puerta::addVuelo(Vuelo* newVuelo) {
    cout << "El vuelo ha sido asignado" << endl;
    act = newVuelo;
    historial.push_back(newVuelo);
}

void Puerta::setV_Asignado(bool Valor) {
    Asignado = Valor;
}

bool Puerta::getV_Asignado() {
    return Asignado;
}

Vuelo* Puerta::getVuelo() {
    return act;
}

void Puerta::setVuelo(Vuelo* newVuelo) {
    act = newVuelo;
    setV_Asignado(true);
}

void Puerta::Clear_Vuelo(){
    Hora_Embarque.SetHora(0, 0);;
    act = nullptr;
    Asignado = false;
}

void Puerta::Show_Info() {
    cout << "ID: " << Id << endl;
    cout << "Ubicacion: " << Ubicacion << endl;
    cout << "Disponibilidad: " << Disponibilidad << endl;
    if (act != nullptr && Asignado) {
        cout << "Hora de Embarque: "; Hora_Embarque.Show_Hora(); cout << endl;
        cout << "Vuelo Asignado: " << endl;
        act->Show_Info(false);
    } 
    else {
        cout << "No hay vuelo asignado." << endl;
    }
    if(!historial.empty()) {
        cout << "Historial de Vuelos:\n" << endl;
        for (int i = 0; i < historial.size(); i++) {
            historial[i]->Show_Info(false); 
        }
    } else {
        cout << "No hay historial de vuelos." << endl;
    }
}

File: puerta.py
from Vuelo import Vuelo  # Importar la clase Vuelo
from Horas import Horas  # Importar la clase Horas

class Puerta:
    def __init__(self, id, ubicacion, disponibilidad):
        self.Id = id
        self.Ubicacion = ubicacion
        self.Disponibilidad = disponibilidad
        self.Hora_Embarque = Horas(0, 0)  # Inicializar Hora_Embarque con valores predeterminados
        self.act = None
        self.Asignado = False
        self.historial = []

    def getId(self):
        return self.Id

    def setId(self, newId):
        self.Id = newId

    def getUbicacion(self):
        return self.Ubicacion

    def setUbicacion(self, newUbicacion):
        self.Ubicacion = newUbicacion

    def getDisponibilidad(self):
        return self.Disponibilidad

    def setDisponibilidad(self, newDisponibilidad):
        self.Disponibilidad = newDisponibilidad

    def getHora_Embarque(self):
        return self.Hora_Embarque

    def setHoraDeEmbarque(self, hora):
        self.Hora_Embarque = hora

    def getHistorial(self):
        tam = len(self.historial)
        Info = 0
        vuelo_Sel = 0
        if tam != 0:
            if tam > 1:
                print("Los vuelos que han sido asignados a esta puerta son:")
                for i in range(tam):
                    print(f"{i + 1}. {self.historial[i].getNo_Identificacion()}.")
                print("Desea conocer más información de alguno de los vuelos?\n1. Si.\n2. No.")
                Info = int(input())
                if Info == 1:
                    print("Sobre qué vuelo desea recibir información? (Indique el número del índice)")
                    vuelo_Sel = int(input())
                    if 1 <= vuelo_Sel <= tam:
                        self.historial[vuelo_Sel - 1].Show_Info(True)
                elif Info == 2:
                    print("Entendido")
                else:
                    print("Valor indefinido")
            else:
                print(f"El único vuelo que ha sido asignado a esta puerta es el vuelo: {self.historial[0].getNo_Identificacion()}")
                print("Desea conocer más información?\n1. Si.\n2. No.")
                Info = int(input())
                if Info == 1:
                    self.historial[0].Show_Info(True)
                elif Info == 2:
                    print("Entendido")
                else:
                    print("Valor indefinido")
        else:
            print("No se ha asignado ningún vuelo a esta puerta")

    def addVuelo(self, newVuelo):
        print("El vuelo ha sido asignado")
        self.act = newVuelo
        self.historial.append(newVuelo)

    def setV_Asignado(self, Valor):
        self.Asignado = Valor

    def getV_Asignado(self):
        return self.Asignado

    def getVuelo(self):
        return self.act

    def setVuelo(self, newVuelo):
        self.act = newVuelo
        self.setV_Asignado(True)

    def Clear_Vuelo(self):
        self.Hora_Embarque.SetHora(0, 0)
        self.act = None
        self.Asignado = False

    def Show_Info(self):
        print(f"ID: {self.Id}")
        print(f"Ubicacion: {self.Ubicacion}")
        print(f"Disponibilidad: {self.Disponibilidad}")
        if self.act is not None and self.Asignado:
            print("Hora de Embarque:", end=" ")
            self.Hora_Embarque.Show_Hora()
            print("\nVuelo Asignado:")
            self.act.Show_Info(False)
        else:
            print("No hay vuelo asignado.")
        if self.historial:
            print("Historial de Vuelos:\n")
            for vuelo in self.historial:
                vuelo.Show_Info(False)
        else:
            print("No hay historial de vuelos.")

